### PR TITLE
Fix TF LED/Longformer attentions computation

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR fixes the test `test_saved_model_with_attentions_output` for TF Longformer and LED that was failing due to an issue in computing some shapes in the attentions.

All the slow tests are now passing 🎉

### CODE REVIEW

File: src/transformers/models/led/modeling_tf_led.py [hunk 1]
```diff
@@ -265,13 +265,26 @@ def call(
...
+                is_index_masked[:, :, None, None],
+                (1, 1, self.num_heads, self.one_sided_attn_window_size * 2 + max_num_global_attn_indices + 1),
+            ),
+            lambda: tf.tile(
+                is_index_masked[:, :, None, None],
+                (1, 1, self.num_heads, self.one_sided_attn_window_size * 2 + 1),
+            ),
+        )
         attn_probs = tf.where(
-            tf.broadcast_to(is_index_masked[:, :, None, None], shape_list(attn_probs)),
```
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_tf_led.py:273]: I don't really understand this change here. The correct shape is given by `attn_probs` -> I don't understand why we cannot just use `shape_list(attn_probs)`? IMO, something like:

```python 
attn_probs = tf.where(
     masked_index,
     tf.zeros(shape_list(masked_index), dtype=attn_probs.dtype
)
``` 

should work, no?

Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_tf_led.py:273]: Also, it's better to make the dtype dependent on the type of `attn_probs`

**Review by patrickvonplaten — COMMENTED** [review]
Thanks for making the test pass! Could you give some more background on why 
a) tf.tile should be used instead of `tf.broadcast_to` & 
b) why we cannot simply use the shape of `attn_probs` since we apply the mask on attn_probs itself? So we know that `shape_list(masked_index) == shape_list(attn_probs)`

**Comment by jplu** [comment]
> a) tf.tile should be used instead of tf.broadcast_to &

There are two reasons for this, the first one is because `broadcast_to` does `reshape` + `tile`, here we don't need to reshape, just `tile` is enough. The second reason is that `broadcast_to` is not compliant with ONNXRuntime.

> b) why we cannot simply use the shape of attn_probs since we apply the mask on attn_probs itself? So we know that shape_list(masked_index) == shape_list(attn_probs)

This part is a bit tricky to explain. The issue here was that `attn_probs` was not always the same shape, if `is_global_attn` is True, then the shape of  `attn_probs` is `[batch_size, seq_len, self.num_heads, self.one_sided_attn_window_size * 2 + max_num_global_attn_indices + 1]`, while if it equals False its shape is `[batch_size, seq_len, self.num_heads, self.one_sided_attn_window_size * 2 + 1]`. Now, because the shape is never potentially the same during the execution when run in graph mode, the pre-computed shape for `attn_probs` by the TF tracing was `[batch_size, seq_len, self.num_heads, variable]`, where `variable` cannot be computed. The consequence of this was that `attn_probs` had never the proper shape at the end and creates a conflict in the `tf.where`. To solve this we had to also create a mask of a fixed shape that depends on `is_global_attn`.

I don't know if it is clear enough or not. Don't hesitate to tell me if there is something you don't get.

File: src/transformers/models/led/modeling_tf_led.py (hunk 1 above)
Comment (jplu) [reply src/transformers/models/led/modeling_tf_led.py:273]: No need here, the default dtype of `tf.zeros` is always float (float16 is AMP is activated, or float32 if not).

**Comment by patrickvonplaten** [comment]
> > a) tf.tile should be used instead of tf.broadcast_to &
> 
> There are two reasons for this, the first one is because `broadcast_to` does `reshape` + `tile`, here we don't need to reshape, just `tile` is enough. The second reason is that `broadcast_to` is not compliant with ONNXRuntime.
> 
> > b) why we cannot simply use the shape of attn_probs since we apply the mask on attn_probs itself? So we know that shape_list(masked_index) == shape_list(attn_probs)
> 
> This part is a bit tricky to explain. The issue here was that `attn_probs` was not always the same shape, if `is_global_attn` is True, then the shape of `attn_probs` is `[batch_size, seq_len, self.num_heads, self.one_sided_attn_window_size * 2 + max_num_global_attn_indices + 1]`, while if it equals False its shape is `[batch_size, seq_len, self.num_heads, self.one_sided_attn_window_size * 2 + 1]`. Now, because the shape is never potentially the same during the execution when run in graph mode, the pre-computed shape for `attn_probs` by the TF tracing was `[batch_size, seq_len, self.num_heads, variable]`, where `variable` cannot be computed. The consequence of this was that `attn_probs` had never the proper shape at the end and creates a conflict in the `tf.where`. To solve this we had to also create a mask of a fixed shape that depends on `is_global_attn`.
> 
> I don't know if it is clear enough or not. Don't hesitate to tell me if there is something you don't get.

Thanks for the explanation - just tried it out and cool to see that your change fixes the test!

**Review by patrickvonplaten — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
If the slow tests pass, LGTM

**Comment by jplu** [comment]
The entire list of slow tests are ok!

**Comment by jplu** [comment]
@sgugger Feel free to merge if it looks ok for you!

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Pull request merged**